src/com/amazon/ion/impl/IonCursorBinary.java

```diff
@@ -178,8 +262,362 @@ class IonCursorBinary implements IonCursor {
+     */
+    private static int logBase2(int value) {
+        return 32 - Integer.numberOfLeadingZeros(value == 0 ? 0 : value - 1);
+    }
+
+    /**
+     * @param value a non-negative number
+     * @return the next power of two greater than or equal to the given number.
+     */
+    static int nextPowerOfTwo(int value) {
+        return (int) Math.pow(2, logBase2(value));
+    }
+
+    /**
+     * Cache of configurations for fixed-sized streams. FIXED_SIZE_CONFIGURATIONS[i] returns a configuration with
+     * buffer size max(8, 2^i). Retrieve a configuration large enough for a given size using
+     * FIXED_SIZE_CONFIGURATIONS(logBase2(size)). Only supports sizes less than or equal to
+     * STANDARD_BUFFER_CONFIGURATION.getInitialBufferSize().
+     */
+    private static final IonBufferConfiguration[] FIXED_SIZE_CONFIGURATIONS;
+
+    static {
+        int maxBufferSizeExponent = logBase2(STANDARD_BUFFER_CONFIGURATION.getInitialBufferSize());
+        FIXED_SIZE_CONFIGURATIONS = new IonBufferConfiguration[maxBufferSizeExponent + 1];
+        for (int i = 0; i <= maxBufferSizeExponent; i++) {
+            // Create a buffer configuration for buffers of size 2^i. The minimum size is 8: the smallest power of two
+            // larger than the minimum buffer size allowed.
+            int size = Math.max(8, (int) Math.pow(2, i));
+            FIXED_SIZE_CONFIGURATIONS[i] = IonBufferConfiguration.Builder.from(STANDARD_BUFFER_CONFIGURATION)
+                .withInitialBufferSize(size)
+                .withMaximumBufferSize(size)
+                .build();
+        }
+    }
+
+    /**
+     * Validates the given configuration.
+     * @param configuration the configuration to validate.
+     * @return the validated configuration.
+     */
+    private static IonBufferConfiguration validate(IonBufferConfiguration configuration) {
+        if (configuration.getInitialBufferSize() < 1) {
+            throw new IllegalArgumentException("Initial buffer size must be at least 1.");
+        }
+        if (configuration.getMaximumBufferSize() < configuration.getInitialBufferSize()) {
+            throw new IllegalArgumentException("Maximum buffer size cannot be less than the initial buffer size.");
+        }
+        return configuration;
+    }
+
+    /**
+     * Constructs a refillable cursor from the given input stream.
+     * @param configuration the configuration to use.
+     * @param alreadyRead the byte array containing the bytes already read (often the IVM).
+     * @param alreadyReadOff the offset into `alreadyRead` at which the first byte that was already read exists.
+     * @param alreadyReadLen the number of bytes already read from `alreadyRead`.
+     */
+    IonCursorBinary(
+        IonBufferConfiguration configuration,
+        InputStream inputStream,
+        byte[] alreadyRead,
+        int alreadyReadOff,
+        int alreadyReadLen
```

Comment:
Consider factoring the handling of these into a different constructor or factory method.

Reply:
These are used *in addition to* the provided InputStream, not as an alternate source. They are used when the caller (in this case the reader builder) has already consumed some bytes from the InputStream in order to determine whether the stream contains binary Ion. We could create a constructor variant that omits these and assumes that the provided InputStream starts at the beginning of the Ion stream, but currently it would only be useful in tests.

Or, did you mean something else? If so, an illustration of what you mean would be helpful.